lib/crawler/keyword/keywords.ex

```diff
@@ -8,28 +8,71 @@ defmodule Crawler.Keyword.Keywords do
   alias Crawler.Keyword.Queries.KeywordQuery
   alias Crawler.Keyword.Schemas.Keyword
   alias Crawler.Repo
+  alias CrawlerWorker.Keyword.CrawlerWorker
 
   def list_keywords(user_id), do: Repo.all(KeywordQuery.user_keywords_query(user_id))
 
+  def get_keyword_by_id(id), do: Repo.get(Keyword, id)
+
   def create_keyword(attrs \\ %{}) do
     %Keyword{}
     |> Keyword.changeset(attrs)
     |> Repo.insert()
   end
 
   def create_keyword_list(keyword_list, user_id) do
-    keyword_list
-    |> build_keyword_params_list(user_id)
-    |> insert_all_keywords()
+    create_keyword_fn = fn k ->
+      case create_keyword(k) do
+        {:ok, %{id: keyword_id}} -> keyword_id
+        # throw  error if a keyword is invalid
+        {:error, changeset} -> throw({:error, changeset})
+      end
+    end
+
+    Repo.transaction(fn ->
+      try do
+        keyword_ids =
+          keyword_list
+          |> build_keyword_params_list(user_id)
+          |> Enum.map(create_keyword_fn)
+
+        enqueue_crawling_job(keyword_ids)
+
+        keyword_ids
+      catch
+        # TODO: handle to show error
+        {:error, _changeset} -> Repo.rollback(:keyword_length_exceeded)
+      end
+    end)
   end
 
-  defp insert_all_keywords(keyword_params_list), do: Repo.insert_all(Keyword, keyword_params_list)
+  def mark_as_in_progress(keyword) do
+    keyword
+    |> Keyword.in_progress_changeset()
+    |> Repo.update()
+  end
 
-  defp build_keyword_params_list(keyword_list, user_id) do
-    now = NaiveDateTime.truncate(NaiveDateTime.utc_now(), :second)
+  def mark_as_completed(keyword, attrs) do
```

Comment:
Let's add tests to all these methods

Reply:
Added in [0eb2154](https://github.com/vnntsu/elixir-ic/pull/39/commits/0eb21542fa957bf89fc373a9e127583cd95edb5d)